File: src/2/3.cpp
#include <iostream>
#include "linked-list.cpp"

void deleteMiddle(Node *node)
{
    if (node == nullptr || node->next == nullptr)
    {
        return;
    }

    Node *next = node->next;
    node->data = next->data;
    node->next = next->next;

    delete next;
}

int main()
{
    LinkedList test = {1, 2, 3, 4, 5, 6, 7};
    deleteMiddle((test.head)->next->next->next);
    test.print();

    return 0;
}

File: src/4/3.js
const { LinkedList } = require("../2/linked-list.js");
const { BSTree } = require("./bst.js");

const listOfDepths = (tree) => {
  const depthMap = new Map();

  const process = (node, depth) => {
    const list = depthMap.get(depth);

    if (list) {
      list.insert(node.data);
    } else {
      depthMap.set(depth, new LinkedList([node.data]));
    }
  };

  const traverse = (node, depth = 0) => {
    if (!node) {
      return;
    }

    process(node, depth);
    traverse(node.left, depth + 1);
    traverse(node.right, depth + 1);
  };

  traverse(tree.head);
  return depthMap;
};

(() => {
  const test = new BSTree([5, 3, 8, 2, 6, 1, 7, 4, 9]);

  console.log(listOfDepths(test));
})();
